components/raftstore/src/store/async_io/write.rs

```diff
@@ -877,39 +901,97 @@ where
         cfg: &Arc<VersionTrack<Config>>,
     ) -> Result<()> {
         let pool_size = cfg.value().store_io_pool_size;
-        for i in 0..pool_size {
-            let tag = format!("store-writer-{}", i);
-            let (tx, rx) = bounded(cfg.value().store_io_notify_capacity);
-            let mut worker = Worker::new(
-                store_id,
-                tag.clone(),
-                raft_engine.clone(),
-                kv_engine.clone(),
-                rx,
-                notifier.clone(),
-                trans.clone(),
-                cfg,
-            );
-            info!("starting store writer {}", i);
-            let t = thread::Builder::new()
-                .name(thd_name!(tag))
-                .spawn_wrapper(move || {
-                    worker.run();
-                })?;
-            self.writers.push(tx);
-            self.handlers.push(t);
+        if pool_size > 0 {
+            self.increase_to(
+                pool_size,
+                StoreWritersContext {
+                    store_id,
+                    notifier: notifier.clone(),
+                    raft_engine,
+                    kv_engine,
+                    transfer: trans.clone(),
+                    cfg: cfg.clone(),
+                },
+            )?;
         }
         Ok(())
     }
 
     pub fn shutdown(&mut self) {
-        assert_eq!(self.writers.len(), self.handlers.len());
-        for (i, handler) in self.handlers.drain(..).enumerate() {
+        let mut handlers = self.handlers.lock();
+        let writers = self.writers.value();
+        assert_eq!(writers.len(), handlers.len());
+        for (i, handler) in handlers.drain(..).enumerate() {
             info!("stopping store writer {}", i);
-            self.writers[i].send(WriteMsg::Shutdown).unwrap();
+            writers[i].send(WriteMsg::Shutdown).unwrap();
             handler.join().unwrap();
         }
     }
+
+    /// Returns the valid size of store writers.
+    pub fn size(&self) -> usize {
+        self.writers.value().capacity()
+    }
+
+    pub fn decrease_to(&mut self, size: usize) -> Result<()> {
+        // Considering that if we released other writers directly, the consistency
+        // of dumping messages by writers will be destroyed when previous writes
+        // haven't been persisted on the waited to be released threads.
+        // So, we just update the capacity of writers to make the concurrent processing
+        // of new messages can be redirected to reserved mailboxes correctly.
```

Comment:
This comment is hard to understand. How about something like this:

"Only update logical capacity but not destroying the workers. So that peers that are still using the writer_id (because there're unpersisted tasks) can proceed. After the peer gets rescheduled, it will use a new writer_id within the new capacity. effectively freeing the worker."